RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,769 @@
+// avoiding any unnecessary resubscriptions
+observable {
+  let! a = foo
+  and! b = bar
+  return a + b
+}
+```
+
+[McBride & Paterson's paper](http://www.staff.city.ac.uk/~ross/papers/Applicative.html) introduces a type very much like F#'s `Result<'T,'TError>` which can be used to stitch together functions and values which might fail, but conveniently accumulating up all of the errors which can then be helpfully presented at once, as opposed to immediately presenting the first error. This allows you to take [Scott Wlaschin's Railway Oriented Programming](https://fsharpforfunandprofit.com/rop/) to the next level by not just bailing out when things go wrong, but automatically collecting up a useful description of all of the different issues that occurred that resulted in the failure.
+
+```fsharp
+// If both reading from the database or the file go wrong, the computation
+// can collect up the errors into a list to helpfully present to the user,
+// rather than just immediately showing the first error and obscuring the
+// second error
+result {
+  let! users = readUsersFromDb()
+  and! birthdays = readUserBirthdaysFromFile(filename)
+  return updateBirthdays users birthdays
+}
+```
+
+[Capriotti & Kaposi's paper](https://paolocapriotti.com/assets/applicative.pdf) introduces an example of creating an command line argument parser, where a single applicative can both statically generate help text for the parser, and dynamically parse options given to an application. [eulerfx](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-309764738) has imagined an F# interpretation of that:
+
+```fsharp
+// One computation expression gives both the behaviour of the parser
+// (in terms of how to parse each element of it, what their defaults should
+// be, etc.) and the information needed to generate its help text
+opt {
+  let! username = Opt("username", (Some ""), Some)
+  and! fullname = Opt("fullname", None, Some)
+  and! id = Opt("id", None, readInt)
+  return User(username, fullname, id)
+}
+```
+
+With all of these examples, we can nest the applicative computation expressions inside other computation expressions to build larger and larger descriptions that cleanly separate the pure computation from its context.
+
+# Detailed Design
+[design]: #detailed-design
+
+This RFC introduces a desugaring for applicative computation expressions, much like that which exists for monads and related constructs.
+
+Example desugaring:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    and! z = baz
+    return x + y + z
+ }
+```
+
+becomes
+
+```fsharp
+ce.Apply(
+    ce.Apply(
+        ce.Apply(
+            ce.Return(
+                (fun x ->
+                    (fun y ->
+                        (fun z ->
+                            x + y + z
+                        )
+                    )
+                )),
+            foo),
+        bar),
+    baz)
+```
+
+## Rules on keywords in applicative computation expressions
```

Comment:
Prior to this section there should be a catalog of the new keywords and what they significance is. You allude to this a bit in this section and with code samples prior to this, but a detailed list of the new keywords will be important before you describe their rules.